File: src/main/java/maifetch/api/MaiTeaClient.java
package maifetch.api;

import maifetch.json.Json;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.math.BigDecimal;
import java.net.HttpURLConnection;
import java.net.URL;
import java.nio.charset.StandardCharsets;
import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

public final class MaiTeaClient {
    static final String DEFAULT_BASE_URL = "https://maitea.app";

    private final String accessToken;
    private final String baseUrl;
    private final Transport transport;

    public MaiTeaClient(String accessToken) {
        this(accessToken, DEFAULT_BASE_URL, new UrlConnectionTransport());
    }

    public MaiTeaClient(String accessToken, String baseUrl, Transport transport) {
        this.accessToken = accessToken == null ? "" : accessToken;
        this.baseUrl = trimTrailingSlash(baseUrl == null || baseUrl.trim().isEmpty() ? DEFAULT_BASE_URL : baseUrl);
        this.transport = transport == null ? new UrlConnectionTransport() : transport;
    }

    public List<Models.Profile> getProfiles() throws IOException {
        Object data = getData("/api/v1/profiles");
        return Models.mapList(data, new Models.Mapper<Models.Profile>() {
            @Override
            public Models.Profile map(Map<String, Object> input) {
                return Models.Profile.fromMap(input);
            }
        });
    }

    public List<Models.TrackInfo> getTracks() throws IOException {
        Object data = getData("/api/v1/tracks");
        return Models.mapList(data, new Models.Mapper<Models.TrackInfo>() {
            @Override
            public Models.TrackInfo map(Map<String, Object> input) {
                return Models.TrackInfo.fromMap(input);
            }
        });
    }

    public Models.Status status() throws IOException {
        Object data = getData("/api/v1/status");
        return Models.Status.fromMap(asMap(data));
    }

    public Pager<Models.Play> getPlays() throws IOException {
        return getPager("/api/v1/plays", playMapper());
    }

    public Pager<Models.Play> getAllPlays() throws IOException {
        return getPager("/api/v1/plays/all", playMapper());
    }

    public Pager<Models.Score> getBestScores() throws IOException {
        return getPager("/api/v1/scores", scoreMapper());
    }

    public Pager<Models.Score> getAllBestScores() throws IOException {
        return getPager("/api/v1/scores/all", scoreMapper());
    }

    <T> Page<T> getPage(String pathOrUrl, Models.Mapper<T> mapper) throws IOException {
        Map<String, Object> root = asMap(request(pathOrUrl));
        Object data = root.get("data");
        List<T> items = Models.mapList(data, mapper);
        Map<String, Object> links = asMap(root.get("links"));
        Map<String, Object> meta = asMap(root.get("meta"));
        return new Page<T>(
                this,
                items,
                stringValue(links.get("first")),
                stringValue(links.get("last")),
                nullableString(links.get("prev")),
                nullableString(links.get("next")),
                intValue(meta.get("current_page")),
                intValue(meta.get("last_page")),
                intValue(meta.get("total"))
        );
    }

    private <T> Pager<T> getPager(String path, Models.Mapper<T> mapper) throws IOException {
        return new Pager<T>(getPage(path, mapper), mapper);
    }

    private Object getData(String path) throws IOException {
        Map<String, Object> root = asMap(request(path));
        return root.get("data");
    }

    private Object request(String pathOrUrl) throws IOException {
        String path = normalizePath(pathOrUrl);
        Map<String, String> headers = new LinkedHashMap<String, String>();
        headers.put("Authorization", "Bearer " + accessToken);
        headers.put("Content-Type", "application/json");
        headers.put("Accept", "application/json");

        Response response = transport.get(baseUrl + path, headers);
        if (response.statusCode < 200 || response.statusCode > 299) {
            throw new IOException("MaiTea request failed with HTTP " + response.statusCode);
        }
        return Json.parse(response.body);
    }

    private String normalizePath(String pathOrUrl) {
        if (pathOrUrl == null || pathOrUrl.trim().isEmpty()) {
            return "/";
        }
        String path = pathOrUrl.trim();
        if (path.startsWith(baseUrl)) {
            path = path.substring(baseUrl.length());
        } else if (path.startsWith(DEFAULT_BASE_URL)) {
            path = path.substring(DEFAULT_BASE_URL.length());
        } else if (path.startsWith("http://") || path.startsWith("https://")) {
            try {
                URL url = new URL(path);
                path = url.getPath() + (url.getQuery() == null ? "" : "?" + url.getQuery());
            } catch (IOException error) {
                throw new IllegalArgumentException("invalid page URL: " + pathOrUrl, error);
            }
        }
        return path.startsWith("/") ? path : "/" + path;
    }

    private static Models.Mapper<Models.Play> playMapper() {
        return new Models.Mapper<Models.Play>() {
            @Override
            public Models.Play map(Map<String, Object> input) {
                return Models.Play.fromMap(input);
            }
        };
    }

    private static Models.Mapper<Models.Score> scoreMapper() {
        return new Models.Mapper<Models.Score>() {
            @Override
            public Models.Score map(Map<String, Object> input) {
                return Models.Score.fromMap(input);
            }
        };
    }

    @SuppressWarnings("unchecked")
    private static Map<String, Object> asMap(Object value) {
        if (value instanceof Map) {
            return (Map<String, Object>) value;
        }
        return Collections.emptyMap();
    }

    private static String stringValue(Object value) {
        String string = nullableString(value);
        return string == null ? "" : string;
    }

    private static String nullableString(Object value) {
        return value == null ? null : String.valueOf(value);
    }

    private static int intValue(Object value) {
        if (value instanceof Number) {
            return ((Number) value).intValue();
        }
        if (value instanceof String && !((String) value).trim().isEmpty()) {
            return new BigDecimal((String) value).intValue();
        }
        return 0;
    }

    private static String trimTrailingSlash(String value) {
        return value.endsWith("/") ? value.substring(0, value.length() - 1) : value;
    }

    public interface Transport {
        Response get(String url, Map<String, String> headers) throws IOException;
    }

    public static final class Response {
        public final int statusCode;
        public final String body;

        public Response(int statusCode, String body) {
            this.statusCode = statusCode;
            this.body = body == null ? "" : body;
        }
    }

    private static final class UrlConnectionTransport implements Transport {
        @Override
        public Response get(String url, Map<String, String> headers) throws IOException {
            HttpURLConnection connection = (HttpURLConnection) new URL(url).openConnection();
            connection.setRequestMethod("GET");
            connection.setConnectTimeout(15000);
            connection.setReadTimeout(30000);
            for (Map.Entry<String, String> header : headers.entrySet()) {
                connection.setRequestProperty(header.getKey(), header.getValue());
            }

            int statusCode = connection.getResponseCode();
            InputStream stream = statusCode >= 400 ? connection.getErrorStream() : connection.getInputStream();
            String body = stream == null ? "" : readAll(stream);
            connection.disconnect();
            return new Response(statusCode, body);
        }

        private static String readAll(InputStream stream) throws IOException {
            try {
                ByteArrayOutputStream output = new ByteArrayOutputStream();
                byte[] buffer = new byte[4096];
                int read;
                while ((read = stream.read(buffer)) != -1) {
                    output.write(buffer, 0, read);
                }
                return new String(output.toByteArray(), StandardCharsets.UTF_8);
            } finally {
                stream.close();
            }
        }
    }
}

File: src/main/kotlin/maifetch/api/MaiTeaClient.kt
package maifetch.api

import maifetch.json.Json
import java.io.ByteArrayOutputStream
import java.io.InputStream
import java.math.BigDecimal
import java.net.HttpURLConnection
import java.net.URL
import java.nio.charset.StandardCharsets

class MaiTeaClient(
    private val accessToken: String,
    baseUrl: String = DEFAULT_BASE_URL,
    private val transport: Transport = UrlConnectionTransport(),
) {
    private val baseUrl: String = trimTrailingSlash(baseUrl.takeUnless { it.isBlank() } ?: DEFAULT_BASE_URL)

    fun getProfiles(): List<Models.Profile> =
        Models.mapList(getData("/api/v1/profiles"), Models.Mapper { Models.Profile.fromMap(it) })

    fun getTracks(): List<Models.TrackInfo> =
        Models.mapList(getData("/api/v1/tracks"), Models.Mapper { Models.TrackInfo.fromMap(it) })

    fun status(): Models.Status = Models.Status.fromMap(asMap(getData("/api/v1/status")))

    fun getPlays(): Pager<Models.Play> = getPager("/api/v1/plays", Models.Mapper { Models.Play.fromMap(it) })

    fun getAllPlays(): Pager<Models.Play> = getPager("/api/v1/plays/all", Models.Mapper { Models.Play.fromMap(it) })

    fun getBestScores(): Pager<Models.Score> = getPager("/api/v1/scores", Models.Mapper { Models.Score.fromMap(it) })

    fun getAllBestScores(): Pager<Models.Score> = getPager("/api/v1/scores/all", Models.Mapper { Models.Score.fromMap(it) })

    internal fun <T> getPage(pathOrUrl: String, mapper: Models.Mapper<T>): Page<T> {
        val root = asMap(request(pathOrUrl))
        val items = Models.mapList(root["data"], mapper)
        val links = asMap(root["links"])
        val meta = asMap(root["meta"])
        return Page(
            client = this,
            data = items,
            first = stringValue(links["first"]),
            last = stringValue(links["last"]),
            previous = nullableString(links["prev"]),
            next = nullableString(links["next"]),
            currentPage = intValue(meta["current_page"]),
            lastPage = intValue(meta["last_page"]),
            total = intValue(meta["total"]),
        )
    }

    private fun <T> getPager(path: String, mapper: Models.Mapper<T>): Pager<T> = Pager(getPage(path, mapper), mapper)

    private fun getData(path: String): Any? = asMap(request(path))["data"]

    private fun request(pathOrUrl: String): Any? {
        val path = normalizePath(pathOrUrl)
        val headers = linkedMapOf(
            "Authorization" to "Bearer $accessToken",
            "Content-Type" to "application/json",
            "Accept" to "application/json",
        )
        val response = transport.get(baseUrl + path, headers)
        if (response.statusCode !in 200..299) {
            throw java.io.IOException("MaiTea request failed with HTTP ${response.statusCode}")
        }
        return Json.parse(response.body)
    }

    private fun normalizePath(pathOrUrl: String?): String {
        if (pathOrUrl.isNullOrBlank()) return "/"
        var path = pathOrUrl.trim()
        path = when {
            path.startsWith(baseUrl) -> path.substring(baseUrl.length)
            path.startsWith(DEFAULT_BASE_URL) -> path.substring(DEFAULT_BASE_URL.length)
            path.startsWith("http://") || path.startsWith("https://") -> {
                val url = URL(path)
                url.path + (url.query?.let { "?$it" } ?: "")
            }
            else -> path
        }
        return if (path.startsWith("/")) path else "/$path"
    }

    interface Transport {
        fun get(url: String, headers: Map<String, String>): Response
    }

    data class Response(
        val statusCode: Int,
        val body: String = "",
    )

    private class UrlConnectionTransport : Transport {
        override fun get(url: String, headers: Map<String, String>): Response {
            val connection = URL(url).openConnection() as HttpURLConnection
            connection.requestMethod = "GET"
            connection.connectTimeout = 15000
            connection.readTimeout = 30000
            headers.forEach { (key, value) -> connection.setRequestProperty(key, value) }

            val statusCode = connection.responseCode
            val stream = if (statusCode >= 400) connection.errorStream else connection.inputStream
            val body = stream?.use { readAll(it) }.orEmpty()
            connection.disconnect()
            return Response(statusCode, body)
        }

        private fun readAll(stream: InputStream): String {
            val output = ByteArrayOutputStream()
            val buffer = ByteArray(4096)
            while (true) {
                val read = stream.read(buffer)
                if (read == -1) break
                output.write(buffer, 0, read)
            }
            return String(output.toByteArray(), StandardCharsets.UTF_8)
        }
    }

    companion object {
        const val DEFAULT_BASE_URL = "https://maitea.app"

        @Suppress("UNCHECKED_CAST")
        private fun asMap(value: Any?): Map<String, Any?> = value as? Map<String, Any?> ?: emptyMap()

        private fun stringValue(value: Any?): String = nullableString(value).orEmpty()

        private fun nullableString(value: Any?): String? = value?.toString()

        private fun intValue(value: Any?): Int =
            when (value) {
                is Number -> value.toInt()
                is String -> value.takeIf { it.isNotBlank() }?.let { BigDecimal(it).toInt() } ?: 0
                else -> 0
            }

        private fun trimTrailingSlash(value: String): String = value.removeSuffix("/")
    }
}
